src/stopwatch.hpp

```diff
@@ -0,0 +1,143 @@
+// bito is free software under the GPLv3; see LICENSE file for details.
+//
+// Simple stopwatch class is for performing runtime speed tests.
+
+#ifndef SRC_STOPWATCH_HPP_
+#define SRC_STOPWATCH_HPP_
+
+#include <stdio.h>
+
+#include <chrono>
+#include <iostream>
+#include <thread>
+#include <vector>
+
+class Stopwatch {
+ public:
+  using clock = std::chrono::high_resolution_clock;
+  using time_point = std::chrono::time_point<clock>;
+  using time_duration = std::chrono::duration<uint16_t, std::nano>;
+
+  enum TimeScale { no_scale, second_scale, millisecond_scale, nanosecond_scale };
+  const static inline double nanosec_per_sec = 1000000000.;
+  const static inline double nanosec_per_millisec = 1000000.;
+
+  // Constructor: option to specify time scale, otherwise defaults to milliseconds.
+  Stopwatch(TimeScale scale = millisecond_scale)
+      : lap_times_(), scale_(scale), total_(0.0), last_subtotal_(0.0) {
+    Start();
+  };
+
+  void Start() { start_time_ = GetCurrentTime(); };
+
+  // Returns time elapsed between current start/stop, and adds that to total time.
+  double Stop() {
+    stop_time_ = GetCurrentTime();
+    latest_subtotal_ = DurationToSeconds(start_time_, stop_time_);
+    total_ += latest_subtotal_;
+    return latest_subtotal_;
+  }
+
+  // Returns time elapsed in last start/stop iteration.
+  double GetLatestSubtotal() { return last_subtotal_; }
+  // Returns total time elapsed from all start/stop iterations.
+  double GetTotal() { return total_; }
+
+  // Returns time elapsed since last start.  Retains no memory of this in Stopwatch.
+  double Elapsed(TimeScale scale = TimeScale::no_scale) {
+    scale = ((scale == TimeScale::no_scale) ? scale_ : scale);
+    stop_time_ = GetCurrentTime();
+    double elapsed_seconds = DurationToSeconds(start_time_, stop_time_, scale);
+    return elapsed_seconds;
+  };
+
+  // Returns the time elapsed since last lap (or start) and appends that to laps.
+  double Lap() {
+    time_point now = GetCurrentTime();
+    size_t num_laps = lap_times_.size();
+    time_point last_lap = ((num_laps == 0) ? start_time_ : lap_times_[num_laps - 1]);
+    lap_times_.push_back(now);
+    double lap_seconds = DurationToSeconds(last_lap, now);
+    lap_seconds_.push_back(lap_seconds);
+    return lap_seconds;
+  };
+  std::vector<double> &GetLaps() { return lap_seconds_; }
+
+  // Clears all total and lap times memory.
+  void Clear() {
+    last_subtotal_ = 0.0;
+    total_ = 0.0;
+    lap_times_.clear();
+    lap_seconds_.clear();
+  }
+
+  // Scale determines what times will be reported in: seconds, milliseconds, or
+  // nanoseconds.
+  void SetScale(TimeScale scale) { scale_ = scale; }
+  TimeScale GetScale() { return scale_; }
+
+  // Puts thread to sleep for given milliseconds.
+  static void Sleep(uint64_t wait_time) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(wait_time));
+  };
+
+  // Stopwatch functionality test.
+  static void UnitTest() {
```

Comment:
Does this actually count as a test?

Reply:
Sorry, it was more of an eyeball test.  I updated it into a doctest without the output.  My issue is that the times aren't deterministic.  We could test for them within a lenient tolerance, but even that test wouldn't guaranteed to pass if the system running it were under load.